go/plugins/googlegenai/gemini.go

```diff
@@ -69,6 +72,265 @@ type EmbedOptions struct {
+	Properties map[string]SchemaProperty `json:"properties,omitempty"`
+	Required   []string                  `json:"required,omitempty"`
+}
+
+type SchemaProperty struct {
+	Type        string                    `json:"type,omitempty"`
+	Description string                    `json:"description,omitempty"`
+	Items       *SchemaProperty           `json:"items,omitempty"`
+	Properties  map[string]SchemaProperty `json:"properties,omitempty"`
+}
+
+type Content struct {
+	Parts []*ContentPart `json:"parts,omitempty"`
+	Role  string         `json:"role,omitempty"`
+}
+
+type ContentPart struct {
+	Text       string      `json:"text,omitempty"`
+	InlineData *InlineData `json:"inlineData,omitempty"`
+}
+
+type InlineData struct {
+	MimeType string `json:"mimeType,omitempty"`
+	Data     string `json:"data,omitempty"`
+}
+
+type RoutingConfig struct {
+	ModelRoutes []*ModelRoute `json:"modelRoutes,omitempty"`
+}
+
+type ModelRoute struct {
+	ModelName string         `json:"modelName,omitempty"`
+	Criteria  *ModelCriteria `json:"criteria,omitempty"`
+}
+
+type ModelCriteria struct {
+	TextLength int32 `json:"textLength,omitempty"`
+	MediaCount int32 `json:"mediaCount,omitempty"`
+}
+
+type ThinkingConfig struct {
+	Enabled bool `json:"enabled,omitempty"`
+}
+
+type HTTPOptions struct {
+	Timeout int `json:"timeout,omitempty"`
+}
+
+// extractConfigFromInput converts any supported config type to GoogleAIConfig
+func extractConfigFromInput(input *ai.ModelRequest) (GeminiConfig, error) {
+	var result GeminiConfig
+	switch config := input.Config.(type) {
+	case GeminiConfig:
+		return config, nil
+	case *GeminiConfig:
+		return *config, nil
+	case ai.GenerationCommonConfig:
+		if config.MaxOutputTokens != 0 {
+			v := int32(config.MaxOutputTokens)
+			result.MaxOutputTokens = &v
+		}
+		result.StopSequences = config.StopSequences
+		if config.Temperature != 0 {
+			v := float32(config.Temperature)
+			result.Temperature = &v
+		}
+		if config.TopK != 0 {
+			v := float32(config.TopK)
+			result.TopK = &v
+		}
+		if config.TopP != 0 {
+			v := float32(config.TopP)
+			result.TopP = &v
+		}
+		result.Version = config.Version
+		return result, nil
+	case *ai.GenerationCommonConfig:
+		if config != nil {
+			if config.MaxOutputTokens != 0 {
+				v := int32(config.MaxOutputTokens)
+				result.MaxOutputTokens = &v
+			}
+			result.StopSequences = config.StopSequences
+			if config.Temperature != 0 {
+				v := float32(config.Temperature)
+				result.Temperature = &v
+			}
+			if config.TopK != 0 {
+				v := float32(config.TopK)
+				result.TopK = &v
+			}
+			if config.TopP != 0 {
+				v := float32(config.TopP)
+				result.TopP = &v
+			}
+			result.Version = config.Version
+		}
+		return result, nil
+	case map[string]any:
+		// Todo: this will silently fail if extra parameters are passed, may want to expose errors
+		if err := mapToStruct(config, &result); err == nil {
+			return result, nil
+		} else {
+			return result, err
```

Comment:
Let's return nil here and elsewhere anytime there's an error. Even though it's nil currently, it's not easy to tell at a glance.